components/scream/scripts/gen_boiler.py

```diff
@@ -0,0 +1,1878 @@
+            end_match = end_regex.match(line)
+            if end_match is not None:
+                expect(active_sub not in db, "Found multiple matches for {}".format(active_sub))
+                expect(len(arg_order) == len(arg_decls),
+                       "Number of decls:\n{}\nDid not match arg list: {}".format(arg_decls, arg_order))
+
+                # we need our decls to be ordered based on arg list order
+                ordered_decls = []
+                for arg in arg_order:
+                    found = False
+                    for arg_decl in arg_decls:
+                        if arg_decl[ARG_NAME] == arg:
+                            ordered_decls.append(arg_decl)
+                            found = True
+                            break
+
+                    expect(found, "Could not find decl for arg {} in\n{}".format(arg, arg_decls))
+
+                db[active_sub] = ordered_decls
+                active_sub = None
+                arg_decls = []
+
+    return db
+
+C_TYPE_MAP = {"real" : "c_real", "integer" : "c_int", "logical" : "c_bool"}
+###############################################################################
+def gen_arg_f90_decl(argtype, intent, dims, names):
+###############################################################################
+    """
+    Generate f90 argument declaration based on the input data
+
+    >>> gen_arg_f90_decl("real", "in", ("10", "150"), ["foo", "bar"])
+    'real(kind=c_real) , intent(in), dimension(10, 150) :: foo, bar'
+    >>> gen_arg_f90_decl("real", "out", ("10", "150"), ["foo", "bar"])
+    'real(kind=c_real) , intent(out), dimension(10, 150) :: foo, bar'
+    >>> gen_arg_f90_decl("logical", "in", None, ["biz", "baz"])
+    'logical(kind=c_bool) , value, intent(in) :: biz, baz'
+    >>> gen_arg_f90_decl("integer", "inout", None, ["barg"])
+    'integer(kind=c_int) , intent(inout) :: barg'
+    >>> gen_arg_f90_decl("integer", "out", None, ["barg"])
+    'integer(kind=c_int) , intent(out) :: barg'
+    """
+    expect(argtype in C_TYPE_MAP, "Unrecognized argtype for C_TYPE_MAP: {}".format(argtype))
+    c_type = C_TYPE_MAP[argtype]
+    value  = ", value" if dims is None and intent == "in" else ""
+    intent_s = ", intent({})".format(intent)
+    dimension_s = ", dimension({})".format(", ".join(dims)) if dims is not None else ""
+    names_s = ", ".join(names)
+    return "{argtype}(kind={c_type}) {value}{intent}{dimension} :: {names}".\
+        format(argtype=argtype, c_type=c_type, value=value, intent=intent_s, dimension=dimension_s, names=names_s)
+
+CXX_TYPE_MAP = {"real" : "Real", "integer" : "Int", "logical" : "bool"}
+###############################################################################
+def get_cxx_type(arg_datum):
+###############################################################################
+    """
+    Based on arg datum, give c++ bridge type
+
+    >>> get_cxx_type(("foo", "real", "in", ("100",)))
```

Comment:
I don't understand the examples. Where would the input strings come from? Would they come from parsing f90 decl lines? Or are the inputs just a random set of strings from which get_cxx_type find the only one (assume there's only one) that matches a known type? I'm guessing it's the former. It might be worth adding a couple of words saying that in the comments, like "the list of strings in the input is assumed to come from the parsing of a f90 type declaration. As such, we _expect only one_ string to match a known F90 type". Or something like that.

Reply:
@bartgol , a core data structure in this program is the sub->args data base where args is a list of tuples (ARG_NAME, ARG_TYPE, ARG_INTENT, ARG_DIMS). These tuples are produced by parse_f90_args. Anywhere you see `arg_datum` it's referring to single one of these tuples. Anywhere you see `arg_data` it's referring to all the tuples for a sub. I will add comments.